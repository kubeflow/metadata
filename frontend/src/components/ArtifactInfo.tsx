/*
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as React from 'react';
import {stylesheet} from 'typestyle';
import {color, commonCss} from '../Css';
import {formatDateString, getArtifactProperty} from '../lib/Utils';
import {MlMetadataArtifact} from '../apis/service';
import {ArtifactProperties, ArtifactCustomProperties} from '../lib/Api';
import {ModelInfo} from './ModelInfo';

export const css = stylesheet({
  artifactInfo: {
    display: 'flex',
    flexDirection: 'row',
    flexWrap: 'wrap',
  },
  field: {
    flexBasis: '300px',
    marginBottom: '32px',
  },
  term: {
    color: color.grey,
    fontSize: '12px',
    letterSpacing: '0.2px',
    lineHeight: '16px',
  },
  value: {
    color: color.secondaryText,
    fontSize: '14px',
    letterSpacing: '0.2px',
    lineHeight: '20px',
  }
});

export interface ArtifactInfoProps {
  typeName: string;
  artifact: MlMetadataArtifact;
}

export class ArtifactInfo extends React.Component<ArtifactInfoProps, {}> {

  public render(): JSX.Element {
    const {artifact} = this.props;
    const createTime = formatDateString(
      getArtifactProperty(artifact, ArtifactProperties.CREATE_TIME) || '');
    return (
      <section>
        <h1 className={commonCss.header}>{this.props.typeName} info</h1>
        <dl className={css.artifactInfo}>
          <div className={css.field}>
            <dt className={css.term}>Version ID</dt>
            <dd className={css.value}>
              {getArtifactProperty(artifact, ArtifactProperties.VERSION)}
            </dd>
          </div>
          <div className={css.field}>
            <dt className={css.term}>Workspace</dt>
            <dd className={css.value}>
              {getArtifactProperty(artifact, ArtifactCustomProperties.WORKSPACE,
                true)}
            </dd>
          </div>
          <div className={css.field}>
            <dt className={css.term}>Run</dt>
            <dd className={css.value}>
              {getArtifactProperty(artifact, ArtifactCustomProperties.RUN,
                true)}
            </dd>
          </div>
          <div className={css.field}>
            <dt className={css.term}>Description</dt>
            <dd className={css.value}>
              {getArtifactProperty(artifact, ArtifactProperties.DESCRIPTION)}
            </dd>
          </div>
          <div className={css.field}>
            <dt className={css.term}>Creation time</dt>
            <dd className={css.value}>{createTime}</dd>
          </div>
          {this.props.typeName === 'Model' ?
            <ModelInfo model={artifact} /> :
            <div className={css.field}>
              <dt className={css.term}>URI</dt>
              <dd className={css.value}>{artifact.uri}</dd>
            </div>
          }
        </dl>
      </section>
    );
  }
}
