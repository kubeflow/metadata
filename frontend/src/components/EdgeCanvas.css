.edgeCanvas {
  position: absolute;
  width: 160px;
  height: 440px;
  border: 0;
  display: inline-block;
  margin-left: 260px;
  margin-top: 74px;
  z-index: -1;
}
.edgeCanvas.reverse {
  margin-left: 0;
  transform: translateX(-100%);
}

.kWKfgJ {
  display: block;
  overflow: visible;
  position: absolute;
}