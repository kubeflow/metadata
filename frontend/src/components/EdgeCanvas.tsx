import * as React from 'react';
import LineChart from 'react-svg-line-chart';
import {classes, stylesheet} from 'typestyle';
import {LineageCardType} from './LineageTypes';
import './LineChart.d.ts';
import {px} from '../Css';
import {CARD_WIDTH, EDGE_WIDTH} from "./LineageCss";

interface EdgeCanvasProps {
  cardArray: number[];
  reverseEdges: boolean;
  type: LineageCardType;
}

export const EdgeCanvas: React.FC<EdgeCanvasProps> = ({cardArray, reverseEdges}) => {
  let viewHeight = 1;

  const cardBodyHeight = 67;
  const cardContainerSpacerHeight = 24;
  const cardTitleHeight = 41;
  const cardTitleBorders = 2;
  const cardOffset = cardContainerSpacerHeight + cardTitleHeight + cardTitleBorders;

  const css = stylesheet({
    edgeCanvas: {
      border: 0,
      display: 'block',
      height: px(440),
      marginLeft: px(CARD_WIDTH),
      marginTop: px(74),
      overflow: 'visible',
      position: 'absolute',
      width: EDGE_WIDTH,
      zIndex: -1,
      '$nest': {
        svg: {
          display: 'block',
          overflow: 'visible',
          position: 'absolute',
        }
      }
    },
    edgeCanvasReverse: {
      marginLeft: 0,
      transform: 'translateX(-100%)',
    },
  });

  const lastNode = reverseEdges ? 'y1' : 'y4';
  const lastNodePositions = {
    y1: 0,
    y4: 0,
  };

  const edgeLines: JSX.Element[] = [];
  cardArray.forEach((rows, i) => {
    for (let j = 0; j < rows; j++) {
      const {y1, y4} = lastNodePositions;
      edgeLines.push(
        <LineChart
          key={`${i}-${j}`}
          data={[
            {x: 0, y: y1},
            {x: 30, y: y1},
            {x: EDGE_WIDTH - 30, y: y4},
            {x: EDGE_WIDTH, y: y4},
          ]}
          areaVisible={false}
          axisVisible={false}
          gridVisible={false}
          labelsVisible={false}
          pathColor={'#BDC1C6'}
          pathVisible={true}
          pathWidth={1}
          pathOpacity={1}
          pointsVisible={false}
          viewBoxHeight={viewHeight}
          viewBoxWidth={EDGE_WIDTH}
          pathSmoothing={0}
        />
      );
      viewHeight += cardBodyHeight;
      lastNodePositions[lastNode] += cardBodyHeight;
    }
    viewHeight += cardOffset;
    lastNodePositions[lastNode] += cardOffset;
  });

  const edgeCanvasClasses = classes(css.edgeCanvas, reverseEdges ? css.edgeCanvasReverse: '');
  return  <div className={edgeCanvasClasses}>{edgeLines}</div>
};
