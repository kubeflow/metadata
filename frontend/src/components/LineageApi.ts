import {ArtifactType, ExecutionType} from "../generated/src/apis/metadata/metadata_store_pb";
import {GetArtifactTypesRequest, GetExecutionTypesRequest} from "../generated/src/apis/metadata/metadata_store_service_pb";
import {MetadataStoreServicePromiseClient} from "../generated/src/apis/metadata/metadata_store_service_grpc_web_pb";

export type ArtifactTypeMap = Map<number, ArtifactType>;
export type ExecutionTypeMap = Map<number, ExecutionType>;

export async function getArtifactTypes(
  metadataStoreService: MetadataStoreServicePromiseClient,
  errorCallback?: (message: string) => void
): Promise<ArtifactTypeMap> {
  const response =
    await metadataStoreService.getArtifactTypes(new GetArtifactTypesRequest());

  if (!response) {
    if (errorCallback) {
      errorCallback('Unable to retrieve Artifact Types, some features may not work.');
    }
    return new Map();
  }

  const artifactTypesMap = new Map<number, ArtifactType>();

  (response!.getArtifactTypesList() || []).forEach((artifactType: ArtifactType) => {
    artifactTypesMap.set(artifactType.getId()!, artifactType);
  });

  return artifactTypesMap;
}

export async function getExecutionTypes(
  metadataStoreService: MetadataStoreServicePromiseClient,
  errorCallback?: (message: string) => void
): Promise<ExecutionTypeMap> {
  const response =
    await metadataStoreService.getExecutionTypes(new GetExecutionTypesRequest());

  if (!response) {
    if (errorCallback) {
      errorCallback('Unable to retrieve Execution Types, some features may not work.');
    }
    return new Map();
  }

  const executionTypesMap = new Map<number, ExecutionType>();

  (response!.getExecutionTypesList() || []).forEach((executionType: ExecutionType) => {
    executionTypesMap.set(executionType.getId()!, executionType);
  });

  return executionTypesMap;
}
