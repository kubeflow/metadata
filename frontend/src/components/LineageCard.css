.cardContainer {
  background: #ffffff;
  border-radius: var(--card-radius);
  width: 260px;
  /* box-shadow: 0 1px #ffffff inset, 0 1px 3px rgba(34, 25, 25, 0.3); */
  border: 1px solid var(--grey-300);
}

.cardContainer .cardTitle {
  height: 40px;
  border-bottom: 1px solid var(--grey-200);
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
}

.cardContainer h3 {
  font-size: 10px;
  color: var(--blue-600);
  letter-spacing: 0.8px;
  text-align: left;
  line-height: 42px;
  padding-left: 20px;
  text-transform: uppercase;
  font-family: "PublicSans-Medium";
}

.cardContainer.addSpacer {
  margin-top: 24px;
}

.cardContainer.target {
  border: 2px solid var(--blue-500);
}

.cardContainer.executions {
  background: #F8FBFF;
  border: 1px solid #CCE4FF;
}
.cardContainer.executions .cardTitle {
  border-bottom: 0;
}
.cardContainer.executions h3 {
  color: #0A3B71;
}

/* Lineage Target Card */
.cardContainer.target {
  border: 2px solid var(--blue-500);
}

/* Lineage Execution Card */
.cardContainer.execution {
  border-radius: var(--card-radius);
  background: #F8FBFF;
  border: 1px solid #CCE4FF;
}

.cardContainer.execution .cardTitle {
  border-top-left-radius: var(--card-radius);
  border-top-right-radius: var(--card-radius);
  border-bottom: 0;
}

.cardContainer.execution .cardTitle h6 {
  color: #0A3B71;
}
