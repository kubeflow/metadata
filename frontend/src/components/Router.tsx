/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import Banner, {BannerProps} from '../components/Banner';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogTitle from '@material-ui/core/DialogTitle';
import Page404 from '../pages/404';
import Snackbar, {SnackbarProps} from '@material-ui/core/Snackbar';
import Toolbar, {ToolbarProps} from './Toolbar';
import {Route, Switch, Redirect, HashRouter} from 'react-router-dom';
import {classes, stylesheet} from 'typestyle';
import {commonCss} from '../Css';
import ArtifactList from '../pages/ArtifactList';
import ArtifactDetails from '../pages/ArtifactDetails';
import ExecutionList from '../pages/ExecutionList';
import SideNav from '../pages/SideNav';

const css = stylesheet({
  dialog: {
    minWidth: 250,
  },
});

export enum RouteParams {
  ARTIFACT_TYPE = 'artifactType',
  EXECUTION_TYPE = 'executionType',
  ID = 'id',
}

export const RoutePage = {
  ARTIFACTS: '/artifacts',
  ARTIFACT_DETAILS: `/artifact_types/:${RouteParams.ARTIFACT_TYPE}+/artifacts/:${RouteParams.ID}`,
  EXECUTIONS: '/executions',
  EXECUTION_DETAILS: `/execution_types/:${RouteParams.EXECUTION_TYPE}+/executions/:${RouteParams.ID}`,
};

export interface DialogProps {
  buttons?: Array<{onClick?: () => any, text: string}>;
  // TODO: This should be generalized to any react component.
  content?: string;
  onClose?: () => any;
  open?: boolean;
  title?: string;
}

interface RouteConfig {
  path: string,
  Component: React.ComponentClass,
  view?: any
}

interface RouteComponentState {
  bannerProps: BannerProps;
  dialogProps: DialogProps;
  snackbarProps: SnackbarProps;
  toolbarProps: ToolbarProps;
}

class Router extends React.Component<{}, RouteComponentState> {

  private routes: RouteConfig[] = [
    {path: RoutePage.ARTIFACTS, Component: ArtifactList},
    {path: RoutePage.ARTIFACT_DETAILS, Component: ArtifactDetails},
    {path: RoutePage.EXECUTIONS, Component: ExecutionList},
    {path: RoutePage.EXECUTION_DETAILS, Component: ArtifactDetails},
  ];

  constructor(props: any) {
    super(props);
    this.state = {
      bannerProps: {},
      dialogProps: {open: false},
      snackbarProps: {autoHideDuration: 5000, open: false},
      toolbarProps: {breadcrumbs: [{displayName: '', href: ''}], actions: [], ...props},
    };
  }

  public render(): JSX.Element {
    const childProps = {
      toolbarProps: this.state.toolbarProps,
      updateBanner: this._updateBanner.bind(this),
      updateDialog: this._updateDialog.bind(this),
      updateSnackbar: this._updateSnackbar.bind(this),
      updateToolbar: this._updateToolbar.bind(this),
    };
    return (
      <HashRouter>
        <div className={commonCss.page}>
          <div className={commonCss.flexGrow}>
          <Route render={({ ...props }) => (<SideNav page={props.location.pathname} {...props} />)} />
            <div className={classes(commonCss.page)}>
              <Route render={({...props}) => (<Toolbar {...this.state.toolbarProps} {...props} />)} />
              {this.state.bannerProps.message
                && <Banner
                  message={this.state.bannerProps.message}
                  mode={this.state.bannerProps.mode}
                  additionalInfo={this.state.bannerProps.additionalInfo}
                  refresh={this.state.bannerProps.refresh} />}
              <Switch>
                <Route exact={true} path={'/'} render={({...props}) => (
                  <Redirect to={RoutePage.ARTIFACTS} {...props} />
                )} />
                {this.routes.map((route, i) => {
                  const {path, Component, ...otherProps} = {...route};
                  return <Route key={i} exact={true} path={path} render={({...props}) => (
                    <Component {...props} {...childProps} {...otherProps} />
                  )} />;
                })}

                {/* 404 */}
                {<Route render={({...props}) => <Page404 {...props} {...childProps} />} />}
              </Switch>

              <Snackbar
                autoHideDuration={this.state.snackbarProps.autoHideDuration}
                message={this.state.snackbarProps.message}
                open={this.state.snackbarProps.open}
                onClose={this._handleSnackbarClose.bind(this)}
              />
            </div>
          </div>

          <Dialog open={this.state.dialogProps.open !== false} classes={{paper: css.dialog}}
            className='dialog' onClose={() => this._handleDialogClosed()}>
            {this.state.dialogProps.title && (
              <DialogTitle> {this.state.dialogProps.title}</DialogTitle>
            )}
            {this.state.dialogProps.content && (
              <DialogContent className={commonCss.prewrap}>
                {this.state.dialogProps.content}
              </DialogContent>
            )}
            {this.state.dialogProps.buttons && (
              <DialogActions>
                {this.state.dialogProps.buttons.map((b, i) =>
                  <Button key={i} onClick={() => this._handleDialogClosed(b.onClick)}
                    className='dialogButton' color='secondary'>
                    {b.text}
                  </Button>)}
              </DialogActions>
            )}
          </Dialog>
        </div>
      </HashRouter>
    );
  }

  private _updateDialog(dialogProps: DialogProps): void {
    // Assuming components will want to open the dialog by defaut.
    if (dialogProps.open === undefined) {
      dialogProps.open = true;
    }
    this.setState({dialogProps});
  }

  private _handleDialogClosed(onClick?: () => void): void {
    this.setState({dialogProps: {open: false}});
    if (onClick) {
      onClick();
    }
    if (this.state.dialogProps.onClose) {
      this.state.dialogProps.onClose();
    }
  }

  private _updateToolbar(newToolbarProps: Partial<ToolbarProps>): void {
    const toolbarProps = Object.assign(this.state.toolbarProps, newToolbarProps);
    this.setState({toolbarProps});
  }

  private _updateBanner(bannerProps: BannerProps): void {
    this.setState({bannerProps});
  }

  private _updateSnackbar(snackbarProps: SnackbarProps): void {
    snackbarProps.autoHideDuration =
      snackbarProps.autoHideDuration || this.state.snackbarProps.autoHideDuration;
    this.setState({snackbarProps});
  }

  private _handleSnackbarClose(): void {
    this.setState({snackbarProps: {open: false, message: ''}});
  }
}

// TODO: loading/error experience until backend is reachable

export default Router;
