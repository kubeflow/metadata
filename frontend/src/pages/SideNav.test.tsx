/*
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';

import SideNav, { css } from './SideNav';
import { LocalStorage } from '../lib/LocalStorage';
import { ReactWrapper, ShallowWrapper, shallow, } from 'enzyme';
import { RouterProps } from 'react-router';
import { RoutePage } from '../components/Router';

const wideWidth = 1000;
const narrowWidth = 200;
const isCollapsed = (tree: ShallowWrapper<any>) =>
  tree.find('WithStyles(IconButton)').hasClass(css.collapsedChevron);
const routerProps: RouterProps = { history: {} as any };

describe('SideNav', () => {
  let tree: ReactWrapper | ShallowWrapper;

  const consoleErrorSpy = jest.spyOn(console, 'error');
  const localStorageHasKeySpy = jest.spyOn(LocalStorage, 'hasKey');
  const localStorageIsCollapsedSpy = jest.spyOn(LocalStorage, 'isNavbarCollapsed');

  beforeEach(() => {
    jest.clearAllMocks();

    consoleErrorSpy.mockImplementation(() => null);

    localStorageHasKeySpy.mockImplementation(() => false);
    localStorageIsCollapsedSpy.mockImplementation(() => false);
  });

  afterEach(async () => {
    // unmount() should be called before resetAllMocks() in case any part of the unmount life cycle
    // depends on mocks/spies
    await tree.unmount();
    jest.resetAllMocks();
    (window as any).innerWidth = wideWidth;
  });

  it('renders expanded state', () => {
    localStorageHasKeySpy.mockImplementationOnce(() => false);
    (window as any).innerWidth = wideWidth;
    tree = shallow(<SideNav page={RoutePage.ARTIFACTS} {...routerProps} />);
    expect(tree).toMatchSnapshot();
  });

  it('renders collapsed state', () => {
    localStorageHasKeySpy.mockImplementationOnce(() => false);
    (window as any).innerWidth = narrowWidth;
    tree = shallow(<SideNav page={RoutePage.ARTIFACTS} {...routerProps} />);
    expect(tree).toMatchSnapshot();
  });

  it('renders artifacts as active page', () => {
    tree = shallow(<SideNav page={RoutePage.ARTIFACTS} {...routerProps} />);
    expect(tree).toMatchSnapshot();
  });

  it('renders artifacts as active when on ArtifactsDetails page', () => {
    tree = shallow(<SideNav page={RoutePage.ARTIFACT_DETAILS} {...routerProps} />);
    expect(tree).toMatchSnapshot();
  });

  it('renders executions as active page', () => {
    tree = shallow(<SideNav page={RoutePage.EXECUTIONS} {...routerProps} />);
    expect(tree).toMatchSnapshot();
  });

  it('renders executions as active when on ExecutionDetails page', () => {
    tree = shallow(<SideNav page={RoutePage.EXECUTION_DETAILS} {...routerProps} />);
    expect(tree).toMatchSnapshot();
  });

  it('collapses if collapse state is true localStorage', () => {
    localStorageIsCollapsedSpy.mockImplementationOnce(() => true);
    localStorageHasKeySpy.mockImplementationOnce(() => true);

    (window as any).innerWidth = wideWidth;
    tree = shallow(<SideNav page={RoutePage.ARTIFACTS} {...routerProps} />);
    expect(isCollapsed(tree)).toBe(true);
  });

  it('expands if collapse state is false in localStorage', () => {
    localStorageIsCollapsedSpy.mockImplementationOnce(() => false);
    localStorageHasKeySpy.mockImplementationOnce(() => true);

    tree = shallow(<SideNav page={RoutePage.ARTIFACTS} {...routerProps} />);
    expect(isCollapsed(tree)).toBe(false);
  });

  it('collapses if no collapse state in localStorage, and window is too narrow', () => {
    localStorageIsCollapsedSpy.mockImplementationOnce(() => false);
    localStorageHasKeySpy.mockImplementationOnce(() => false);

    (window as any).innerWidth = narrowWidth;
    tree = shallow(<SideNav page={RoutePage.ARTIFACTS} {...routerProps} />);
    expect(isCollapsed(tree)).toBe(true);
  });

  it('expands if no collapse state in localStorage, and window is wide', () => {
    localStorageIsCollapsedSpy.mockImplementationOnce(() => false);
    localStorageHasKeySpy.mockImplementationOnce(() => false);

    (window as any).innerWidth = wideWidth;
    tree = shallow(<SideNav page={RoutePage.ARTIFACTS} {...routerProps} />);
    expect(isCollapsed(tree)).toBe(false);
  });

  it('collapses if no collapse state in localStorage, and window goes from wide to narrow', () => {
    localStorageIsCollapsedSpy.mockImplementationOnce(() => false);
    localStorageHasKeySpy.mockImplementationOnce(() => false);

    (window as any).innerWidth = wideWidth;
    tree = shallow(<SideNav page={RoutePage.ARTIFACTS} {...routerProps} />);
    expect(isCollapsed(tree)).toBe(false);

    (window as any).innerWidth = narrowWidth;
    const resizeEvent = new Event('resize');
    window.dispatchEvent(resizeEvent);
    expect(isCollapsed(tree)).toBe(true);
  });

  it('expands if no collapse state in localStorage, and window goes from narrow to wide', () => {
    localStorageIsCollapsedSpy.mockImplementationOnce(() => false);
    localStorageHasKeySpy.mockImplementationOnce(() => false);

    (window as any).innerWidth = narrowWidth;
    tree = shallow(<SideNav page={RoutePage.ARTIFACTS} {...routerProps} />);
    expect(isCollapsed(tree)).toBe(true);

    (window as any).innerWidth = wideWidth;
    const resizeEvent = new Event('resize');
    window.dispatchEvent(resizeEvent);
    expect(isCollapsed(tree)).toBe(false);
  });

  it('saves state in localStorage if chevron is clicked', () => {
    localStorageIsCollapsedSpy.mockImplementationOnce(() => false);
    localStorageHasKeySpy.mockImplementationOnce(() => false);
    const spy = jest.spyOn(LocalStorage, 'saveNavbarCollapsed');

    (window as any).innerWidth = narrowWidth;
    tree = shallow(<SideNav page={RoutePage.ARTIFACTS} {...routerProps} />);
    expect(isCollapsed(tree)).toBe(true);

    tree.find('WithStyles(IconButton)').simulate('click');
    expect(spy).toHaveBeenCalledWith(false);
  });

  it('does not collapse if collapse state is saved in localStorage, and window resizes', () => {
    localStorageIsCollapsedSpy.mockImplementationOnce(() => false);
    localStorageHasKeySpy.mockImplementationOnce(() => true);

    (window as any).innerWidth = wideWidth;
    tree = shallow(<SideNav page={RoutePage.ARTIFACTS} {...routerProps} />);
    expect(isCollapsed(tree)).toBe(false);

    (window as any).innerWidth = narrowWidth;
    const resizeEvent = new Event('resize');
    window.dispatchEvent(resizeEvent);
    expect(isCollapsed(tree)).toBe(false);
  });
});
