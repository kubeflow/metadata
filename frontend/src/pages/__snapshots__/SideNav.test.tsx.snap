// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideNav renders artifacts as active page 1`] = `
<div
  className="root flexColumn noShrink"
  id="sideNav"
>
  <div
    style={
      Object {
        "flexGrow": 1,
      }
    }
  >
    <div
      className="indicator"
    />
    <WithStyles(Tooltip)
      disableFocusListener={true}
      disableHoverListener={true}
      disableTouchListener={true}
      enterDelay={300}
      placement="right-start"
      title="Artifacts List"
    >
      <Link
        className="unstyled"
        id="artifactsBtn"
        replace={false}
        to="/artifacts"
      >
        <WithStyles(Button)
          className="button active"
        >
          <pure(BubbleChartIcon) />
          <span
            className="label"
          >
            Artifacts
          </span>
        </WithStyles(Button)>
      </Link>
    </WithStyles(Tooltip)>
    <div
      className="indicator indicatorHidden"
    />
    <WithStyles(Tooltip)
      disableFocusListener={true}
      disableHoverListener={true}
      disableTouchListener={true}
      enterDelay={300}
      placement="right-start"
      title="Execution List"
    >
      <Link
        className="unstyled"
        id="executionsBtn"
        replace={false}
        to="/executions"
      >
        <WithStyles(Button)
          className="button"
        >
          <pure(PlayArrowIcon) />
          <span
            className="label"
          >
            Executions
          </span>
        </WithStyles(Button)>
      </Link>
    </WithStyles(Tooltip)>
    <hr
      className="separator"
    />
    <WithStyles(IconButton)
      className="chevron"
      onClick={[Function]}
    >
      <pure(ChevronLeftIcon) />
    </WithStyles(IconButton)>
  </div>
</div>
`;

exports[`SideNav renders artifacts as active when on ArtifactsDetails page 1`] = `
<div
  className="root flexColumn noShrink"
  id="sideNav"
>
  <div
    style={
      Object {
        "flexGrow": 1,
      }
    }
  >
    <div
      className="indicator indicatorHidden"
    />
    <WithStyles(Tooltip)
      disableFocusListener={true}
      disableHoverListener={true}
      disableTouchListener={true}
      enterDelay={300}
      placement="right-start"
      title="Artifacts List"
    >
      <Link
        className="unstyled"
        id="artifactsBtn"
        replace={false}
        to="/artifacts"
      >
        <WithStyles(Button)
          className="button active"
        >
          <pure(BubbleChartIcon) />
          <span
            className="label"
          >
            Artifacts
          </span>
        </WithStyles(Button)>
      </Link>
    </WithStyles(Tooltip)>
    <div
      className="indicator indicatorHidden"
    />
    <WithStyles(Tooltip)
      disableFocusListener={true}
      disableHoverListener={true}
      disableTouchListener={true}
      enterDelay={300}
      placement="right-start"
      title="Execution List"
    >
      <Link
        className="unstyled"
        id="executionsBtn"
        replace={false}
        to="/executions"
      >
        <WithStyles(Button)
          className="button"
        >
          <pure(PlayArrowIcon) />
          <span
            className="label"
          >
            Executions
          </span>
        </WithStyles(Button)>
      </Link>
    </WithStyles(Tooltip)>
    <hr
      className="separator"
    />
    <WithStyles(IconButton)
      className="chevron"
      onClick={[Function]}
    >
      <pure(ChevronLeftIcon) />
    </WithStyles(IconButton)>
  </div>
</div>
`;

exports[`SideNav renders collapsed state 1`] = `
<div
  className="root flexColumn noShrink collapsedRoot"
  id="sideNav"
>
  <div
    style={
      Object {
        "flexGrow": 1,
      }
    }
  >
    <div
      className="indicator"
    />
    <WithStyles(Tooltip)
      disableFocusListener={false}
      disableHoverListener={false}
      disableTouchListener={false}
      enterDelay={300}
      placement="right-start"
      title="Artifacts List"
    >
      <Link
        className="unstyled"
        id="artifactsBtn"
        replace={false}
        to="/artifacts"
      >
        <WithStyles(Button)
          className="button active collapsedButton"
        >
          <pure(BubbleChartIcon) />
          <span
            className="collapsedLabel label"
          >
            Artifacts
          </span>
        </WithStyles(Button)>
      </Link>
    </WithStyles(Tooltip)>
    <div
      className="indicator indicatorHidden"
    />
    <WithStyles(Tooltip)
      disableFocusListener={false}
      disableHoverListener={false}
      disableTouchListener={false}
      enterDelay={300}
      placement="right-start"
      title="Execution List"
    >
      <Link
        className="unstyled"
        id="executionsBtn"
        replace={false}
        to="/executions"
      >
        <WithStyles(Button)
          className="button collapsedButton"
        >
          <pure(PlayArrowIcon) />
          <span
            className="collapsedLabel label"
          >
            Executions
          </span>
        </WithStyles(Button)>
      </Link>
    </WithStyles(Tooltip)>
    <hr
      className="separator collapsedSeparator"
    />
    <WithStyles(IconButton)
      className="chevron collapsedChevron"
      onClick={[Function]}
    >
      <pure(ChevronLeftIcon) />
    </WithStyles(IconButton)>
  </div>
</div>
`;

exports[`SideNav renders executions as active page 1`] = `
<div
  className="root flexColumn noShrink"
  id="sideNav"
>
  <div
    style={
      Object {
        "flexGrow": 1,
      }
    }
  >
    <div
      className="indicator indicatorHidden"
    />
    <WithStyles(Tooltip)
      disableFocusListener={true}
      disableHoverListener={true}
      disableTouchListener={true}
      enterDelay={300}
      placement="right-start"
      title="Artifacts List"
    >
      <Link
        className="unstyled"
        id="artifactsBtn"
        replace={false}
        to="/artifacts"
      >
        <WithStyles(Button)
          className="button"
        >
          <pure(BubbleChartIcon) />
          <span
            className="label"
          >
            Artifacts
          </span>
        </WithStyles(Button)>
      </Link>
    </WithStyles(Tooltip)>
    <div
      className="indicator"
    />
    <WithStyles(Tooltip)
      disableFocusListener={true}
      disableHoverListener={true}
      disableTouchListener={true}
      enterDelay={300}
      placement="right-start"
      title="Execution List"
    >
      <Link
        className="unstyled"
        id="executionsBtn"
        replace={false}
        to="/executions"
      >
        <WithStyles(Button)
          className="button active"
        >
          <pure(PlayArrowIcon) />
          <span
            className="label"
          >
            Executions
          </span>
        </WithStyles(Button)>
      </Link>
    </WithStyles(Tooltip)>
    <hr
      className="separator"
    />
    <WithStyles(IconButton)
      className="chevron"
      onClick={[Function]}
    >
      <pure(ChevronLeftIcon) />
    </WithStyles(IconButton)>
  </div>
</div>
`;

exports[`SideNav renders executions as active when on ExecutionDetails page 1`] = `
<div
  className="root flexColumn noShrink"
  id="sideNav"
>
  <div
    style={
      Object {
        "flexGrow": 1,
      }
    }
  >
    <div
      className="indicator indicatorHidden"
    />
    <WithStyles(Tooltip)
      disableFocusListener={true}
      disableHoverListener={true}
      disableTouchListener={true}
      enterDelay={300}
      placement="right-start"
      title="Artifacts List"
    >
      <Link
        className="unstyled"
        id="artifactsBtn"
        replace={false}
        to="/artifacts"
      >
        <WithStyles(Button)
          className="button"
        >
          <pure(BubbleChartIcon) />
          <span
            className="label"
          >
            Artifacts
          </span>
        </WithStyles(Button)>
      </Link>
    </WithStyles(Tooltip)>
    <div
      className="indicator"
    />
    <WithStyles(Tooltip)
      disableFocusListener={true}
      disableHoverListener={true}
      disableTouchListener={true}
      enterDelay={300}
      placement="right-start"
      title="Execution List"
    >
      <Link
        className="unstyled"
        id="executionsBtn"
        replace={false}
        to="/executions"
      >
        <WithStyles(Button)
          className="button active"
        >
          <pure(PlayArrowIcon) />
          <span
            className="label"
          >
            Executions
          </span>
        </WithStyles(Button)>
      </Link>
    </WithStyles(Tooltip)>
    <hr
      className="separator"
    />
    <WithStyles(IconButton)
      className="chevron"
      onClick={[Function]}
    >
      <pure(ChevronLeftIcon) />
    </WithStyles(IconButton)>
  </div>
</div>
`;

exports[`SideNav renders expanded state 1`] = `
<div
  className="root flexColumn noShrink"
  id="sideNav"
>
  <div
    style={
      Object {
        "flexGrow": 1,
      }
    }
  >
    <div
      className="indicator"
    />
    <WithStyles(Tooltip)
      disableFocusListener={true}
      disableHoverListener={true}
      disableTouchListener={true}
      enterDelay={300}
      placement="right-start"
      title="Artifacts List"
    >
      <Link
        className="unstyled"
        id="artifactsBtn"
        replace={false}
        to="/artifacts"
      >
        <WithStyles(Button)
          className="button active"
        >
          <pure(BubbleChartIcon) />
          <span
            className="label"
          >
            Artifacts
          </span>
        </WithStyles(Button)>
      </Link>
    </WithStyles(Tooltip)>
    <div
      className="indicator indicatorHidden"
    />
    <WithStyles(Tooltip)
      disableFocusListener={true}
      disableHoverListener={true}
      disableTouchListener={true}
      enterDelay={300}
      placement="right-start"
      title="Execution List"
    >
      <Link
        className="unstyled"
        id="executionsBtn"
        replace={false}
        to="/executions"
      >
        <WithStyles(Button)
          className="button"
        >
          <pure(PlayArrowIcon) />
          <span
            className="label"
          >
            Executions
          </span>
        </WithStyles(Button)>
      </Link>
    </WithStyles(Tooltip)>
    <hr
      className="separator"
    />
    <WithStyles(IconButton)
      className="chevron"
      onClick={[Function]}
    >
      <pure(ChevronLeftIcon) />
    </WithStyles(IconButton)>
  </div>
</div>
`;
